adr/2025_06_13-document-transforms.md

```diff
@@ -0,0 +1,65 @@
+# Document Transforms
+
+A core part of the BlockNote API is the ability to make changes to the document, using BlockNote's core design of blocks, this is much easier to do programmatically than with other editors.
+
+We've done pretty well with our existing API, but there are a few things that could be improved.
+
+Referencing content within the document is either awkward or non-existent. Right now we essentially really only have an API for referencing blocks by their id with no further level of granularity.
+
+## Locations
+
+[Looking at Slate](https://docs.slatejs.org/concepts/03-locations) (highly recommend reading the docs), they have the concept of a `Location` which is a way to reference a specific point in the document, but it does not have to be so specific as positions, it has levels of granularity.
+
+This gives us a unified way to reference content within the document, allowing for much more granular editing. Take a look at the `Location.ts` file for more details around this.
```

Comment:
Do we want to necessarily use the `Location` concept? I like the granularity but it feels like we could achieve something similar by just using block/inline content ID references. I.e.:

- Low granularity: Point to the start/end of a block by its ID.
- Mid granularity: Point to the start/end of inline content by its ID.
- High granularity: Point to the start of inline content with an offset by its ID.

I feel like Slate's `Location` is necessary due to the fact that you can only reference nodes by their positions. Since we set IDs, I think we should make use of that since imo it'll result in a simpler API.